homeassistant/components/omnilogic/switch.py

```diff
@@ -0,0 +1,256 @@
+"""Platform for switch integration."""
+import time
+
+from omnilogic import OmniLogicException
+import voluptuous as vol
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.helpers import config_validation as cv, entity_platform
+
+from .common import OmniLogicEntity, OmniLogicUpdateCoordinator, check_guard
+from .const import COORDINATOR, DOMAIN, PUMP_TYPES
+
+SERVICE_SET_SPEED = "set_pump_speed"
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the light platform."""
+
+    coordinator = hass.data[DOMAIN][entry.entry_id][COORDINATOR]
+    entities = []
+
+    for item_id, item in coordinator.data.items():
+        id_len = len(item_id)
+        item_kind = item_id[-2]
+        entity_settings = SWITCH_TYPES.get((id_len, item_kind))
+
+        if not entity_settings:
+            continue
+
+        for entity_setting in entity_settings:
+            for state_key, entity_class in entity_setting["entity_classes"].items():
+                if check_guard(state_key, item, entity_setting):
+                    continue
+
+                entity = entity_class(
+                    coordinator=coordinator,
+                    state_key=state_key,
+                    name=entity_setting["name"],
+                    kind=entity_setting["kind"],
+                    item_id=item_id,
+                    icon=entity_setting["icon"],
+                )
+
+                entities.append(entity)
+
+    async_add_entities(entities)
+
+    # register service
+    platform = entity_platform.current_platform.get()
+
+    platform.async_register_entity_service(
+        SERVICE_SET_SPEED,
+        {vol.Required("speed"): cv.positive_int},
+        "async_set_speed",
+    )
+
+
+class OmniLogicSwitch(OmniLogicEntity, SwitchEntity):
+    """Define an Omnilogic Switch entity."""
+
+    def __init__(
+        self,
+        coordinator: OmniLogicUpdateCoordinator,
+        kind: str,
+        name: str,
+        icon: str,
+        item_id: tuple,
+        state_key: str,
+    ):
+        """Initialize Entities."""
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            item_id=item_id,
+            icon=icon,
+        )
+
+        self._state_key = state_key
+        self._state = None
+        self._last_action = 0
+        self._state_delay = 30
+
+    @property
+    def is_on(self):
+        """Return the on/off state of the switch."""
+        if self._last_action < (time.time() - self._state_delay):
+            self._state = int(self.coordinator.data[self._item_id][self._state_key])
+
+            if self._state == 7:
+                self._state = 0
+
+        return self._state
+
+
+class OmniLogicRelayControl(OmniLogicSwitch):
+    """Define the OmniLogic Relay entity."""
+
+    async def async_turn_on(self, **kwargs):
+        """Turn on the relay."""
+        self._state = True
```

Comment:
Instead of optimistically setting the state, can the actual state be checked after the `set_relay_valve` call?

Reply:
As mentioned, the actual state isn't reflected in the telemetry from the Omnilogic API for several (up to 60) seconds after triggering the `set_relay_valve` call so we are setting the state optimistically and then causing a delay in state reporting to allow the telemetry to catch up.

Reply:
OK, please add a comment explaining that.